File: java/BinarySearchTree.java
public class BinarySearchTree {
    // private Node root=null;
    BinarySearchTree() {
        Node root = null;
    }

    public Node insert(Node root, int d, double gpa) {
        if(root == null)
            return new Node(d, gpa);
        else if(root.rollno >= d)
        {
            root.left = insert(root.left, d, gpa);
        }
        else 
            root.right = insert(root.right, d, gpa);
        return root;
    } 

    public boolean search(Node root, int d) {
        if(root==null)
            return false;
        if(d==root.rollno)
            return true;
        else if(d<=root.rollno)
            search(root.left, d);
        else
            search(root.right, d);
        return true;
    }

    public Node findmin(Node root) {
        while(root.left!= null)
            root = root.left;
        return root;
    }

    public Node deleteNode(Node root, int d)
    {
        if(root==null)
            return root;
        else if(d<root.rollno)
            deleteNode(root.left, d);
        else if(d>root.rollno)
            deleteNode(root.right, d);
        else
        {
            if(root.left == null)
                return root.right;
            else if(root.right==null)
                return root.left;
            else
            {
                Node temp = findmin(root.right);
                root.rollno = temp.rollno;
                root.cgpa = temp.cgpa;
                root.right = deleteNode(root.right, root.rollno);
            }
        }
        return root;
    }

    public void inorder(Node root) {
        if(root==null)
            return;
        inorder(root.left);
        System.out.println(root.rollno + " " + root.cgpa);
        inorder(root.right);
    }

    public void preorder(Node root) {
        if(root == null)
            return;
        System.out.println(root.rollno + " " + root.cgpa);
        preorder(root.left);
        preorder(root.right);
        
    }

    public static void main(String[] args) {
        // Node root;
        BinarySearchTree tree = new BinarySearchTree();
        Node root = null;
        root = tree.insert(root, 4, 9.2);
        root = tree.insert(root, 3, 9.4);
        root = tree.insert(root, 2, 9.5);
        root = tree.insert(root, 1, 9.1);
        root = tree.insert(root, 5, 9.0);
        root = tree.insert(root, 6, 9.8);
        root = tree.insert(root, 7, 8.6);
        tree.inorder(root);
        System.out.println();
        tree.preorder(root);
        // tree.insert(2);
    }

}

File: APC/Algorithms & Data Structrures/binary_search_tree.cpp
#include <iostream>
#include <stdbool.h>

using namespace std;

typedef struct Node
{
    int data;
    Node *left;
    Node *right;
} Node;

Node *createNewNode(int data)
{
    Node *nn = new Node();
    nn->data = data;
    nn->left = NULL;
    nn->right = NULL;
    return nn;
}

Node *insert(Node *root, int d)
{
    if (root == NULL)
        return createNewNode(d);
    else if (root->data <= d)
    {
        root->right = insert(root->right, d);
    }
    else
        root->left = insert(root->left, d);

    return root;
}
Node *findmin(Node *root)
{
    // root = root->left;
    while (root->left)
        root = root->left;
    return root;
}

Node *deleteNode(Node *root, int d)
{
    if (root == NULL)
        return root;
    else if (d < root->data)
        root->left = deleteNode(root->left, d);
    else if (d > root->data)
        root->right = deleteNode(root->right, d);
    else
    {
        if (root->left == NULL && root->right == NULL)
        {
            delete root;
            root = NULL;
        }
        else if (root->left == NULL)
        {
            Node *temp = root;
            root = root->right;
            delete temp;
        }
        else if (root->right == NULL)
        {
            Node *temp = root;
            root = root->left;
            delete temp;
        }
        else
        {
            Node *temp = findmin(root->right);
            root->data = temp->data;
            root->right = deleteNode(root->right, temp->data);
        }
    }
    return root;
}

bool search(Node *root, int d)
{
    if (root == NULL)
        return false;
    if (d == root->data)
        return true;
    else if (d <= root->data)
        search(root->left, d);
    else
        search(root->right, d);
}

void inorder(Node *root)
{
    if (root == NULL)
        return;
    inorder(root->left);
    cout << root->data << " ";
    inorder(root->right);
}

void preorder(Node *root)
{
    if (root == NULL)
        return;
    cout << root->data << " ";
    preorder(root->left);
    preorder(root->right);
}

int main()
{
    Node *root = NULL;
    root = insert(root, 5);
    root = insert(root, 4);
    root = insert(root, 6);
    root = insert(root, 3);
    root = insert(root, 7);
    root = insert(root, 2);
    root = insert(root, 8);
    root = insert(root, 1);
    root = insert(root, 9);
    deleteNode(root, 1);
    inorder(root);
    cout << endl;
    preorder(root);
    cout << endl;
    cout << search(root, 1) << endl;
    return 0;
}